From 69dd556466277ee3dcbd90f63b90f01f33175ce9 Mon Sep 17 00:00:00 2001
From: Po Liu <Po.Liu@nxp.com>
Date: Thu, 23 Mar 2017 22:43:20 +0800
Subject: [PATCH]  add sja1105-netconf command to create xml file for
 edit-config in netconf

xml registers value get by registers:
sja1105-netconf config save xxx.xml

Signed-off-by: Po Liu <Po.Liu@nxp.com>
---
 Makefile                            |  7 ++++-
 src/config/xml/write/config-write.c | 60 +++++++++++++++++++++++++++++++++++++
 2 files changed, 66 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index bb1c7e7..0fa9868 100644
--- a/Makefile
+++ b/Makefile
@@ -43,17 +43,22 @@ MANPAGES = docs/man/sja1105-tool.1 \
            docs/man/sja1105-tool-config-format.5 \
            docs/man/sja1105-conf.5
 SJA1105  = sja1105-tool
+SJA1105_NETCONF = sja1105-netconf
 
-build: $(SJA1105)
+build: $(SJA1105) $(SJA1105_NETCONF)
 
 $(SJA1105): $(DEPS)
 	$(CC) $(OBJ) -o $@ $(LDFLAGS)
 
+$(SJA1105_NETCONF): $(SRC)
+	$(CC) $(CFLAGS) -DNETCONF $(filter %.c, $^) -o $@ $(LDFLAGS)
+
 %.o: %.c
 	$(CC) $(CFLAGS) -c $^ -o $@
 
 clean:
 	rm -f $(SJA1105) $(OBJ)
+	rm -f $(SJA1105) $(SJA1105_NETCONF) $(MANPAGES) $(OBJ)
 
 man: $(MANPAGES)
 
diff --git a/src/config/xml/write/config-write.c b/src/config/xml/write/config-write.c
index 7d03e4d..399d3be 100644
--- a/src/config/xml/write/config-write.c
+++ b/src/config/xml/write/config-write.c
@@ -44,8 +44,18 @@ int xml_write_field(xmlTextWriterPtr writer, char *field, uint64_t value)
 {
 	char print_buf[MAX_LINE_SIZE];
 
+#if defined(NETCONF)
+       if (!strcmp(field, "index"))
+               snprintf(print_buf, MAX_LINE_SIZE, "%lld", value);
+       else
+#endif
 	snprintf(print_buf, MAX_LINE_SIZE, "0x%" PRIX64, value);
+#ifndef NETCONF
 	return xmlTextWriterWriteAttribute(writer, BAD_CAST field, BAD_CAST print_buf);
+#else
+	return xmlTextWriterWriteElement(writer, BAD_CAST field, BAD_CAST print_buf);
+#endif
+
 }
 
 int xml_write_array(xmlTextWriterPtr writer, char *field, uint64_t *values, int count)
@@ -53,7 +63,11 @@ int xml_write_array(xmlTextWriterPtr writer, char *field, uint64_t *values, int
 	char print_buf[MAX_LINE_SIZE];
 
 	print_array(print_buf, values, count);
+#ifndef NETCONF
 	return xmlTextWriterWriteAttribute(writer, BAD_CAST field, BAD_CAST print_buf);
+#else
+	return xmlTextWriterWriteElement(writer, BAD_CAST field, BAD_CAST print_buf);
+#endif
 }
 
 int write_config_tables(xmlTextWriterPtr writer, struct sja1105_config *config)
@@ -102,6 +116,7 @@ int write_config_tables(xmlTextWriterPtr writer, struct sja1105_config *config)
 		retagging_table_write,
 		xmii_mode_parameters_table_write,
 	};
+#ifndef	NETCONF
 	uint64_t entry_counts[] = {
 		config->schedule_count,
 		config->schedule_entry_points_count,
@@ -124,15 +139,44 @@ int write_config_tables(xmlTextWriterPtr writer, struct sja1105_config *config)
 		0,
 		config->xmii_params_count,
 	};
+#else
+	uint64_t entry_counts[] = {
+	1,
+	1,
+	0,
+	0,
+	0,
+	1,
+	1,
+	1,
+	1,
+	1,
+	1,
+	1,
+	0,
+	1,
+	1,
+	0,
+	0,
+	1,
+	0,
+	1,
+	};
+#endif
+
 	int rc = 0;
 	unsigned int i;
 
 	for (i = 0; i < ARRAY_SIZE(options); i++) {
 		if (entry_counts[i]) {
+#ifndef NETCONF
 			rc |= xmlTextWriterStartElement(writer, BAD_CAST "table");
 			rc |= xmlTextWriterWriteAttribute(writer,
 			      BAD_CAST "name",
 			      BAD_CAST options[i]);
+#else
+			rc |= xmlTextWriterStartElement(writer, BAD_CAST options[i]);
+#endif
 			rc |= next_write_config_table[i](writer, config);
 			rc |= xmlTextWriterEndElement(writer);
 			if (rc < 0) {
@@ -163,7 +207,9 @@ int sja1105_config_write_to_xml(char *filename, struct sja1105_config *config)
 		goto out;
 	}
 	rc |= xmlTextWriterSetIndent(writer, 1);
+#ifndef NETCONF
 	rc |= xmlTextWriterSetIndentString(writer, (const xmlChar*) "\t");
+#endif
 	if (rc < 0) {
 		loge("could not set xml indentation");
 		goto out;
@@ -172,10 +218,24 @@ int sja1105_config_write_to_xml(char *filename, struct sja1105_config *config)
 	if (rc < 0) {
 		goto out;
 	}
+#ifndef NETCONF
 	rc = xmlTextWriterStartElement(writer, BAD_CAST "config");
 	if (rc < 0) {
 		goto out;
 	}
+#else
+	rc = xmlTextWriterStartElement(writer, BAD_CAST "sja1105");
+	if (rc < 0) {
+		goto out;
+	}
+	rc |= xmlTextWriterWriteAttribute(writer,
+			      BAD_CAST "xmlns",
+			      BAD_CAST "http://nxp.com/sja1105");
+	if (rc < 0) {
+		goto out;
+	}
+#endif
+
 	rc = write_config_tables(writer, config);
 	if (rc < 0) {
 		loge("could not write config tables");
-- 
1.8.3.1

